### PR TITLE
Fix promoted enum array self-casts

### LINKED ISSUE #7991: Useless promoted cast to enum causes illegal cast error
Normally, a cast of a type to itself does nothing. Furthermore, a cast can be promoted when provided with an array element. But, a useless promoted cast to an enum type causes an internal compiler error when it should do nothing.

``` chapel
enum myenum {
  ONE=1,
  TWO=2,
  THREE=3
}
use myenum;

var A:[1..3] myenum = [ONE, TWO, THREE];
writeln(A);

var B = A:int;
writeln(B);

var C = B:int; // this useless cast is promoted, makes C = B.
writeln(C);

var D = C:myenum; // demonstrating cast to enum can be promoted
writeln(D);

var F = A:myenum; // results in compilation error
writeln(F);
```

causes error
```
useless-promoted-cast.chpl:20: error: illegal cast from [domain(1,int(64),false)] myenum to myenum
```

### PR DESCRIPTION
**Pull request by bradcray**
This change fixes issue #7991 by providing an identity self-cast from
an enum value to itself (this is similar to how promoted self-casts on
other types like arrays of integers work as well).  It adds the test from
that issue as a guard against backsliding in the future.

resolves #7991.

### CODE REVIEW

**Comment by bradcray** [comment]
@mppf: Could I ask you to review this since it resolves an issue that you filed and touches code that you've edited recently?  Also: do you remember what caused you to run into this?  (i.e., does it result in cleanup in practical use cases that we care about / demos / benchmarks / etc.?)

File: modules/internal/ChapelBase.chpl [hunk 1]
```diff
@@ -1112,6 +1112,8 @@ module ChapelBase {
     return x: int: bool;
   // _cast(type t:integral, x:enumerated)
   // is generated for each enum in buildDefaultFunctions
+  inline proc _cast(type t:enumerated, x:enumerated) where x.type == t
```
[the remark this answers is not in the document]
Comment (bradcray) [reply modules/internal/ChapelBase.chpl]: I didn't try writing it that way, but am pretty sure we'd run into issue #9310 (which is what I got the last time I tried patterns like this... once bitten twice shy...).  The effect would be that we'd support casting `color`s to `size`s (but then return a `color` anyway, I'd guess?).  Most of the current ChapelBase.chpl generic operators on enumerated types are written this way for this reason.

**Comment by bradcray** [comment]
> It seems strange to me that we need a `_cast` overload for this case

It looks to me like we support them for other cases.  E.g.,

```chapel
  inline proc _cast(type t:integral, x:integral)
    ...
  inline proc _cast(type t:chpl_anyreal, x:chpl_anyreal)
    ...
  inline proc _cast(type t:chpl_anycomplex, x: chpl_anycomplex)
    ...
```

Is there a specific reason you think enums should be different in this respect?

**Comment by bradcray** [comment]
> It looks to me like we support them for other cases.

To verify my assumption, I used the `--explain-call` flag on some simple programs that do self-casts.  Neither the integer nor the enumerated self-cast overloads are used for scalar-to-scalar self-casts.  But the integer self-cast in `ChapelBase.chpl` that I pointed to above is what's used in the promoted int-array-cast-to-int case (and the new enumerated self-cast is what's used for the code from this issue and filed in this PR).

**Comment by mppf** [comment]
> Is there a specific reason you think enums should be different in this respect?

I didn't remember the other cases had these as well. That makes me more confident in this change, thanks.

**Pull request merged**